opt: support SHOW TRACE

RaduBerinde opened the pull request:
#### sql: move explain ResultColumns to sqlbase

We are duplicating the information about the result columns returned
by EXPLAIN statements (in sql and in opt/optbuilder). This change
moves the definitions to sqlbase and uses them in both cases.

Release note: None

#### opt: support SHOW TRACE

Adding optimizer support for SHOW TRACE statements.

We use two operators, one for `SHOW TRACE FOR <query>` and one for
`SHOW TRACE FOR SESSION`. They both use the same `ShowTraceOpDef` for
the private.

We don't add non-explain execbuilder tests because statement tracing
doesn't work with the TestEngine implementation. This will be
addressed in a subsequent PR.

Release note: None

RaduBerinde committed:
opt: support SHOW TRACE

Adding optimizer support for SHOW TRACE statements.

We use two operators, one for `SHOW TRACE FOR <query>` and one for
`SHOW TRACE FOR SESSION`. They both use the same `ShowTraceOpDef` for
the private.

We don't add non-explain execbuilder tests because statement tracing
doesn't work with the TestEngine implementation. This will be
addressed in a subsequent PR.

Release note: None

RaduBerinde committed:
sql: move explain ResultColumns to sqlbase

We are duplicating the information about the result columns returned
by EXPLAIN statements (in sql and in opt/optbuilder). This change
moves the definitions to sqlbase and uses them in both cases.

Release note: None

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, nice work

---

Review status: 0 of 25 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/opt/memo/private_defs.go, line 154 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25851#-LDCexIj4Nhcph0OklKV:-LDCexIk2Y_qnZQFpVkr:ben1msj) ([raw file](https://github.com/cockroachdb/cockroach/blob/02f71bdfc329e37dc14d3c5d140a0f8c2ad8efc4/pkg/sql/opt/memo/private_defs.go#L154)):*
> ```Go
> 	Type tree.ShowTraceType
> 
> 	Compact bool
> ```

NIT: comment would be useful here

---

*[pkg/sql/opt/ops/relational.opt, line 322 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25851#-LDCfGvT7XiLYIyNVQb3:-LDCfGvT7XiLYIyNVQb4:b-1uwton) ([raw file](https://github.com/cockroachdb/cockroach/blob/02f71bdfc329e37dc14d3c5d140a0f8c2ad8efc4/pkg/sql/opt/ops/relational.opt#L322)):*
> ```
> }
> 
> # ShowTrace is an operator which returns information about the execution plan of
> ```

NIT: ShowTrace => Explain

Also, I'd get rid of the "is an operator which" here and below (and above, actually, lol).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25851#-:-LDCgHrf6anSuEiifpvU:b-qyf0xd)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: disable alter_table logictest for opt

Removing the opt configs for the `alter_table` logic test, which now
fails some SHOW KV TRACE statements because we choose a different
index. I will split it out in a separate change.

Release note: None

RaduBerinde commented:
bors r+
